test/async_worker.cc

```diff
@@ -95,12 +98,65 @@ class TestWorkerNoCallback : public AsyncWorker {
   bool _succeed;
 };
 
+class EchoWorker : public AsyncWorker {
+ public:
+  EchoWorker(Function& cb, std::string& echo) : AsyncWorker(cb), echo(echo) {}
+  ~EchoWorker() {}
+
+  void Execute() override {
+    // Simulate cpu heavy task
+    std::this_thread::sleep_for(std::chrono::milliseconds(30));
+  }
+
+  void OnOK() override {
+    HandleScope scope(Env());
+    Callback().Call({Env().Null(), String::New(Env(), echo)});
+  }
+
+ private:
+  std::string echo;
+};
+
+class CancelWorker : public AsyncWorker {
+ public:
+  CancelWorker(Function& cb) : AsyncWorker(cb) {}
+  ~CancelWorker() {}
+
+  static void DoWork(const CallbackInfo& info) {
+    Function cb = info[0].As<Function>();
+    std::string echo = info[1].As<String>();
+    for (int i = 1; i < MAX_CANCEL_THREADS; i++) {
```

Comment:
Maybe we can get the cpu number with JS API and fill the worker threads with that number of workers? In that case, the test can be more stable.

With JS specified worker filling number, we can also test that `cancelWorker->Cancel()` throws if the cancel worker has been scheduled.

Reply:
Hey Legendecas!  Thank you for the review. By cpu numbers do you mean the size of the Libuv's thread pool? I tested this by setting `UV_THREADPOOL_SIZE` to smaller numbers such as 2 or 3 and it seems to work reliably in that the CancelWorker is able to cancel the queue'd task. 

Also, I am running into some difficulties testing that `Cancel()` throws if the worker has started the task. Initially, I just passed the value 0 as the threadNum to queue a task for the Cancel worker to cancel, expecting an error to be thrown. For some reason, this exception is only thrown once by `Cancel()`. 

After doing some investigation, it looks like we run this test using `common.runTest` where we run the test function using 4 different node-gyp configs in rapid succession. Could it be that by the second time the `test` function is invoked, the thread pool is busy and cancellation is successful? If that's the case is there a way where we could check to make sure the threads are idle before running the test suite? Thanks.

Reply:
Looks like we already have a similar test case in core (https://github.com/nodejs/node/blob/main/test/node-api/test_async/test_async.c#L140) and it has been running in a quite stable way for a long time. If that's the case, I'm fine with the current pattern as it is the simplest way to test the API.

Yeah, if the test case is mixed up with other test suites, we can not guarantee that the threadpool is ready to execute the work immediately. But I believe we can still facilitate mutexes and conditional variables to notify the JS thread when the Worker is started to execute. However, this increases the complexity of this test case and I assume it is fine to do it in a follow up PR.